ivadomed/utils.py

```diff
@@ -236,14 +239,34 @@ def structurewise_uncertainty(fname_lst, fname_hard, fname_unc_vox, fname_out):
         nib_im = nib.load(fname)
         data_im = nib_im.get_fdata()
         data_lst.append(data_im)
-        data_im_l, _ = label(data_im, structure=bin_struct)
+        data_im_l = []
+        for i in range(data_im.shape[-1]):
+            data_im_tmp, _ = label(data_im[..., i], structure=bin_struct)
+            data_im_l.append(data_im_tmp)
         data_l_lst.append(data_im_l)
         del nib_im
+    data_l_lst = np.array(data_l_lst)
+    # channel went first due to the 'append' function
+    # Befor transpose we are supposed to have MC_simulation X Channel X Height X Width X Depth
+    # After transpose we expect to have MC_simulation X Height X Width X Depth X Channel
+    # this is wanted because the network output has a shape of Height X Width X Depth X Channel
+    # So the rest of the code is designed to use something like this
+    if len(data_l_lst.shape) == 4:
+        data_l_lst = np.transpose(data_l_lst, (0, 2, 3, 4, 1))
+    elif len(data_l_lst.shape) == 3:
+        data_l_lst = np.transpose(data_l_lst, (0, 2, 3, 1))
 
     # loop across all structures of data_hard_l
     for i_l in range(1, n_l + 1):
```

Comment:
What if the different channel have a different n_l? Here n_l is the one of the last processed channel (see L225)

Reply:
I see what you mean, I will need to loop over the n_l I guess.

Reply:
Another solution would be to check if the n_l are actually different before looping over them, don't know if it really slows the program down.

Reply:
But then: I am a bit lost a line 287, because it won't have the same shape as data_l_lst[i_mc]